WallsAndHoles/tilemap.cpp

```diff
@@ -1,19 +1,17 @@
 #include "tilemap.h"
 
 TileMap::TileMap(QSize mapSize,
-                 const SharedTileTemplate defaultTileTemplate,
                  QObject *parent)
     : QObject(parent)
     , mMapSize(mapSize)
-    , mDefaultTileTemplate(defaultTileTemplate)
 {
-    mMap = (QSharedPointer<Tile> **)malloc(sizeof(QSharedPointer<Tile> *) * mapSize.width());
-    for (int x = 0; x < mapSize.width(); ++x) {
-        mMap[x] = (QSharedPointer<Tile> *)malloc(sizeof(QSharedPointer<Tile>) * mapSize.height());
+    mMap = (Tile ***)malloc(sizeof(Tile **) * mMapSize.width());
```

Comment:
malloc() and free() are C-style coding. I recommend using

mMap = new Tile**[ mMapSize.width() ];
mMap[x] = new Tile*[ mMapSize.height() ];  for all x
mMap[x][y] = new Tile(...);  for all x, y

and then don't forget to add to your destructor:

delete mMap[x][y]; for all x,y
delete [] mMap[x]; for all x
delete [] mMap;


Even better: make a class for 2D arrays that uses vectors to implement them. This moves all the messy vector<vector<>> stuff outside the Tile code, allows easy resizability, and gets rid of any chance of a memory leak (as long as you use a vector<vector<shared pointer>>)

Reply:
right now your destructor does not free mMap completely